src/cryptsetup/cryptsetup-generator.c

```diff
@@ -106,11 +167,30 @@ static int create_disk(
                 "DefaultDependencies=no\n"
                 "Conflicts=umount.target\n"
                 "IgnoreOnIsolate=true\n"
-                "After=%s\n",
+                "After=%s\n\n",
                 netdev ? "remote-fs-pre.target" : "cryptsetup-pre.target");
 
+        if (keydev) {
+                _cleanup_free_ char *unit = NULL, *mount = NULL, *p = NULL;
+
+                r = generate_keydev_mount(name, keydev, &unit, &mount);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to generate keydev mount unit: %m");
+
+                p = prefix_root(mount, password_escaped);
+                if (!p)
+                        return log_oom();
+                free_and_replace(password_escaped, p);
+
+                /* Cleanup the keydev mount after the attach */
+                fprintf(f,
+                        "[Service]\n"
+                        "ExecStartPost=" UMOUNT_PATH " %s\n\n", mount);
+        }
+
         if (!nofail)
                 fprintf(f,
+                        "[Unit]\n"
```

Comment:
This feels brittle. In fact, several section below write dependencies without having a [Unit] in them. I think it's better if you split the if(keydev) up just like tmp or swp are.

Reply:
Yupp, that is actually a very good point.